source/pool/contracts/Pool.sol

```diff
@@ -0,0 +1,161 @@
+  event Bulk(bytes32[] roots, address account, address token, uint256 amount);
+
+  /**
+   * @notice Structs
+   */
+  struct Claim {
+    bytes32 root;
+    uint256 score;
+    bytes32[] proof;
+  }
+
+  /**
+   * @notice Constructor
+   * @param _scale uint256
+   * @param _max uint256
+   */
+  constructor(uint256 _scale, uint256 _max) public {
+    scale = _scale;
+    max = _max;
+  }
+
+  /**
+   * @notice Enables claims on the merkle tree of a set of scores
+   * @param root bytes32
+   */
+  function enable(bytes32 root) external onlyOwner {
+    require(roots[root] == false, "ROOT_EXISTS");
+    roots[root] = true;
+    emit Enable(root);
+  }
+
+  /**
+   * @notice Single claim of tokens to be transferred to msg.sender
+   * @param root bytes32
+   * @param score uint256
+   * @param proof bytes32[]
+   * @param token address
+   */
+  function single(
+    bytes32 root,
+    uint256 score,
+    bytes32[] memory proof,
+    address token
+  ) public {
+    require(roots[root], "ROOT_NOT_ENABLED");
+    require(!claimed[root][msg.sender], "CLAIM_INVALID");
+    require(verify(msg.sender, root, score, proof), "PROOF_INVALID");
+    claimed[root][msg.sender] = true;
+
+    uint256 amount = getOutput(score, token);
+    IERC20(token).transfer(msg.sender, amount);
+    emit Single(root, msg.sender, token, amount);
+  }
+
+  /**
+   * @notice Bulk claim of tokens to be transferred to msg.sender
+   * @param claims Claim[]
+   * @param token address
+   */
+  function bulk(Claim[] memory claims, address token) public {
+    require(claims.length > 0, "NO_CLAIMS_PROVIDED");
+    uint256 totalScore = 0;
+    bytes32[] memory rootList = new bytes32[](claims.length);
+    Claim memory claim;
+    for (uint256 i = 0; i < claims.length; i++) {
+      claim = claims[i];
+      require(roots[claim.root], "ROOT_NOT_ENABLED");
+      require(!claimed[claim.root][msg.sender], "CLAIM_INVALID");
+      require(
+        verify(msg.sender, claim.root, claim.score, claim.proof),
+        "PROOF_INVALID"
+      );
+      totalScore = totalScore + claim.score;
+      claimed[claim.root][msg.sender] = true;
+    }
+    uint256 amount = getOutput(totalScore, token);
+    IERC20(token).transfer(msg.sender, amount);
+    emit Bulk(rootList, msg.sender, token, amount);
+  }
+
+  /**
+   * @notice Get output amount for an input score
+   * @param score uint256
+   * @param token address
+   */
+  function getOutput(uint256 score, address token)
+    public
+    view
+    returns (uint256 amount)
+  {
+    return
+      (max *
+        ((score * IERC20(token).balanceOf(address(this))) /
+          ((10**scale) + score))) / 100;
+  }
+
+  /**
+   * @notice Verify a claim proof
+   * @param participant address
+   * @param root bytes32
+   * @param score uint256
+   * @param proof bytes32[]
+   */
+  function verify(
+    address participant,
+    bytes32 root,
+    uint256 score,
+    bytes32[] memory proof
```

Comment:
I'm sure OZ thought of this, but one concern I have (despite the complexity growing by log(n)) is that the proof is too large to be computed over. I don't see is hitting this limit any time soon.

Reply:
maybe if the merkle tree ends up being too big, it can be divided into 2 or more trees and provide multiple roots

Reply:
Yeah, actually that's an interesting point. @dmosites is each vote it's own merkle?

Reply:
Figure with 10K leaves (that's a lot of voters) the proof lands around 10 hashes.